### PR TITLE
Replicate DeepestLSTMTinyPilotNet model from Tensorflow to PyTorch framework

### LINKED ISSUE #37: Replicating DeepestLSTMTinyPilotNet model from Tensorflow to PyTorch framework
The current model is implemented in Tensorflow here: https://github.com/JdeRobot/DeepLearningStudio/blob/main/Formula1-FollowLine/tensorflow/DeepestLSTMTinyPilotNet/utils/deepest_lstm_tinypilotnet.py.

It would be a good feature to support it in Pytorch and utilize it there.

### PR DESCRIPTION
**Pull request by nik1806**
Fixes #37

### CODE REVIEW

**Comment by sergiopaniego** [comment]
Does the model complete different circuits? Could you attach a video and share the weights so we can reproduce the results easily?

**Comment by nik1806** [comment]
Hello, In the following drive you can find the trained weights and a small video. https://drive.google.com/drive/folders/1rhx4q3nfPT2EGPZ66pgKlalZp6lyoFx3?usp=sharing

The performance is not as good as a PilotNet, probably because of the new brain file I created for the PyTorch (considering the TensorFlow code from 'brains/f1/brain_keras-lstm.py') for simulation in BehavorMetrics is not perfect and needs fine-tuning. Could you please help me with some resources or tips or implementation to understand the logic and fine-tune the brain files?

**Comment by sergiopaniego** [comment]
Hi!

Are you training with sequences (many-to-one) or with the same approach as the PilotNet model (one image as input each time). If the case is the second one, the brain on Behavior Metrics should be exactly the same as for PilotNet. The brain that you pointed is a draft one, from some previous iterations I tried. This is the brain that should be used (https://github.com/JdeRobot/BehaviorMetrics/blob/noetic-devel/behavior_metrics/brains/f1/brain_f1_keras.py)

When you talk about performance, are you talking about the MSE and MAE losses in train-validation? the performance of the brain on the circuits (it fails on some curves)?

**Comment by nik1806** [comment]
Hello!
I have run new simulations and now the model can easily complete simple_circuit. The model, video, and brain files are available in the drive: https://drive.google.com/drive/folders/1rhx4q3nfPT2EGPZ66pgKlalZp6lyoFx3?usp=sharing
I have used sequences (many-to-one) for training similar to what was done with the TensorFlow version using similar processing.
Please let me know about the next steps, and whether you need some improvements or updates before finalizing.

**Comment by sergiopaniego** [comment]
Nice job!
Next step is creating a PR on https://github.com/JdeRobot/BehaviorMetrics with the new brain file for torch with sequences since both PRs are complementary

**Comment by nik1806** [comment]
Thank you. I included all supporting code for simulation with this PR https://github.com/JdeRobot/BehaviorMetrics/pull/335.

**Comment by sergiopaniego** [comment]
Hi! I added some feedback on Behavior Metrics's PR that you should review. 
Could you give me details about the experimental loss that you obtained when training the models, comparing the PilotNet model vs this implementation of DeepestLSTM. Was the loss on Deepest close to the one obtained when training PilotNet?

**Comment by nik1806** [comment]
Hi! Thank you for providing feedback on Behavior Metric's PR. I have the changes accordingly in the new commit. 

About the loss during training, when I trained PilotNet, the model converges at a decent phase to a value around 0.5. With DeepestLSTM, the convergence was faster (at an earlier epoch), then some minor decrease to a value around 0.04. The loss values are close to each other (although since models are different direct value comparison is hard, both model converges which is a positive sign). 

With a hyperparameter search in training, we might be able to squeeze out more performance, but I am not sure if it is required now since it is resource-intensive.
